Proper guest management in backoffice

Hlavtox opened the pull request:
**Could be improved**
- 🚧 Group selectors have absolutely no selector and targeting select all checkbox is not very clean, it could disable other inputs by accident. More precise selector would be better.

| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | See below
| Type?             | new feature
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes #22645 and #25355
| How to test?      | See below
| Possible impacts? | No

### Description
- This PR allows you to create a guest customer from backoffice.
- There is a new field when creating a customer, allows you to quickly create a guest. 
- If guest is checked, password field, enable field and group fields are greyed out to behave the same as in FO. And they get changed to proper groups so the merchant knows what will happen after saving.
- Once you save this customer, the same logic as in front office is processed - the guest has a random password and his group is default guest group from PS configuration. No matter what was sent in the form.
- The field is not shown for existing customers. If converting a customer, we have separate functions for that. And we don't want people to be able to mess up a registered customer by changing him back to guest. We can work on this in separate PR.
- This field shows up only if guest checkout is enabled, because Customer class does not allow to create a guest if guest checkout is disabled.
![guest](https://user-images.githubusercontent.com/6097524/155132322-7ce4a258-cd8f-4d73-a5d9-bca9cd2e9b86.png)

### Technical changes
- I added some JS that, depending on guest field, disables and clears password input, sets groups, sets them disabled, disables "is enabled" switch. They are overriden anyway in FormHandler, but so that the visual experience matches what is done later internally. And for security.
- I pass default groups to template so when switching guest ON/OFF I can re-check those 3 default groups. Their IDs are configurable so you can't use just 1,2,3.
- Added a parameter coming to Customer form which shows the guest field only when creating a customer.
- Added a `FormEvents::PRE_SUBMIT` listener that removes constraints from password field if the is_guest was sent and is true. Standard formCloner class is used for this - I got inspired in other places in the code. Added a missing NotBlank constraint, because it was validated in the ValueObject only.
- Adjusted the command in FormHandler before sending it further into the handler. It sets a random password, guest group, sets enabled to true.

### Video
https://user-images.githubusercontent.com/6097524/219753944-37e4603c-fa60-4b1c-a3d1-cf8c7126afa3.mp4

https://user-images.githubusercontent.com/6097524/219767560-09ebcf92-0a66-4b32-9d6b-ac5f813698b9.mp4

### Manual tests
 - Create guest through front office and back office and check that they are the same data-wise.
 - Make a registered customer A first@prestashop.com. Check that you can create a new customer with the same email ONLY if you are creating a guest.

### Automatic behavior tests
- Everything is already automatically tested - implemented in previous PRs.
- OK to create a guest customer with existing email.
- FAIL to create a duplicate customer with existing email
- OK to change guest's email to already existing email.
- FAIL to change registered customer's email to already existing email.

### Preparations for this PR
I have extracted the majority of this PR to separate PRs, so it's easier to review.
- https://github.com/PrestaShop/PrestaShop/pull/31375
- https://github.com/PrestaShop/PrestaShop/pull/31386
- https://github.com/PrestaShop/PrestaShop/pull/31363
- https://github.com/PrestaShop/PrestaShop/pull/25353
- https://github.com/PrestaShop/PrestaShop/pull/31461
- https://github.com/PrestaShop/PrestaShop/pull/31462

Hlavtox committed:
Implement guest management into backoffice

boubkerbribri commented on tests/Integration/Behaviour/Features/Scenario/Customer/customer_management.feature:
```diff
@@ -151,3 +151,37 @@ Feature: Customer Management
+      | lastName  | Customer                    |
+      | email     | customereleven@prestashop.com     |
+      | password  | PrestaShopForever1_!        |
+    And I attempt to edit customer "CUST-10" and I change the following properties:
```
```suggestion
    And I attempt to edit customer "CUST-10" with the following properties:
```

boubkerbribri commented on tests/Integration/Behaviour/Features/Scenario/Customer/customer_management.feature:
```diff
@@ -151,3 +151,37 @@ Feature: Customer Management
+      | password  | PrestaShopForever1_!        |
+    And I edit customer "CUST-8" and I change the following properties:
+      | email | customernine@prestashop.com |
+    Then if I query customer "CUST-8" I should get a Customer with properties:
```
This step is not defined either, the closest thing i could find is https://github.com/PrestaShop/PrestaShop/blob/develop/tests/Integration/Behaviour/Features/Context/CustomerManagerFeatureContext.php#L227

Hlavtox commented:
@atomiix @jolelievre Guys, do you think you could spare a minute and help me with this PR? 🙏 The two remaining things are on the top of PR description. 🥺🥺🥺

kpodemski commented:
@Hlavtox you need help with these points:

> TODO - if the field is checked
> 🚧 Disable password input visually and do not require it. Password is ignored anyway in case of guest creation.
> 🚧 Disable group_ids and default_group_id inputs and do not require it. Groups are ignored anyway in case of guest creation.

?

Hlavtox commented:
@kpodemski Yeeees, exactly. :-)

Julievrz commented on src/PrestaShopBundle/Form/Admin/Sell/Customer/CustomerType.php:
```diff
@@ -107,6 +107,17 @@ public function __construct(
+        if ($options['show_guest_field'] === true) {
+            $builder
+                ->add('is_guest', SwitchType::class, [
+                    'label' => $this->trans('Register as guest', 'Admin.Global'),
```
Hi :) I think we should avoid having the word "register" and "guest" in the same sentence. In my opinion, it's kind of confusing. We could use "Guest checkout" as a label. I think this expression is quite known in ecommerce, so I'm not sure it requires a help text. What could be nice would be to display this feature just under the password field and have the password field grayed out when this feature is enabled? What do you think?

Hlavtox replied:
Hello Julie, I am not against some other label, but I don't like `Guest checkout`, because we are not in a 'checkout' when creating the customer in backoffice. Can you try to come up with some other options? 😉

Julievrz replied:
Yes, sorry, I read too quickly and I thought we were on the back-office order page.  What about "Guest" only? And for the help text:

'Continue without creating a customer account"

Hlavtox replied:
What about "Create a guest"?

`Continue without creating a customer account` - Well, you create a customer account, but of a guest type. I think the help text should mean something like: `This will create a guest account with no password and a default guest group. Customer won't be able to log in with this account, unless it's converted to normal customer account. This creates the same type of customer as if the customer does guest checkout in the front office.` ➡️ Can you make a help text out of this? 😄

Julievrz replied:
Ok, well, I think the misunderstanding comes from the fact that, in my opinion, a guest is not a customer. And I wouldn't talk about a "guest account". It's confusing, as being a guests means that you don't have an account in the FO (but there's one "account" in the BO)

But I understand your point. 

Here's another suggestion:

The user won't have a password to sign in to your store unless you turn their account into a customer one.

(The merchant will have to do the modification, right?)

Hlavtox replied:
- They can convert their guest account on order confirmation or guest tracking page.
- They are customers. The only difference is is_guest flag being true + not being able to log in with it. They are in the same table, in the same listing, the same entity. :-)

Hlavtox replied:
We can just call it:
`Create guest account`
`This will create a single use account with no password. User won't be able to sign in with it, unless he converts it to a regular account.`

Julievrz replied:
Ok, thanks for your explanation. The following suggestion should do it then. :) 
Label: "Create a guest" as you suggested
Help text: "The user won't have a password to sign in to your store unless they create an account."

Hlavtox replied:
@Julievrz They don't create an account, they convert it from guest to customer type.

eshraw commented:
Hello @Hlavtox , thanks for the PR and your work ! I think this is a nice addition for one time customers created in the BO.

After some investigation on the topic of guests, I have a couple of things I would like to discuss since you have a good grasp on the subject : 
- The concept of guest account is problematic from a semantic point of view. As a guest you don't expect to be able to transform a "guest account" into a "customer account" because you don't expect to have an account. If you decided to check-out as a guest you don't expect data persistence nor the possibility to mutate a "guest account", you would expect to have to sign up some way or another.
- We lack consistency between the pages with options to convert a guest to a customer in the FO : 
   - In the check-out the guest is suggested to sign-up with a sign-up form (so far so good, seems logical).
   - In the guest order tracking page the guest is suggested to transform a "guest account" into a "customer account" by setting a password --> this differs from the check-out convert attempt and promotes this concept of "guest account" that we would probably like to avoid since it doesn't make much sense (semantically speaking).
- The wordings (related to both others) : 
   -  While I understand that technically, a guest in PrestaShop is a customer account with is_guest == true, semantically and functionally we need to steer away from that because it creates a situation in which it implies that we are storing guest checkout data into an account they didn't ask for. I stand by @Julievrz when she propose that we use "_Create a guest_".
   - In the BO, we can use the wording "_A guest won't have a password unless they create an account or you convert them to a customer's account_". If it's done in the FO, by the customer, we need to have it look like they are creating an account.
   - In the FO, we need to be more consistent on guest-to-customer conversion, I believe we can be smart with our wordings here and use something along the lines of "_Re-use your order informations to create an account_" + password input on both pages allowing this conversion. That way the user is under the impression that they are creating an account while, technically, we are only converting a guest into a customer. (This is probably something to be addressed in another PR/issue, but I believe it's related to your PR in some capacity)
- The last point I'd like to discuss is how you would display guests in the customer's list ? As you mentionned, guests are customers with is_guest == true, thus they appear in the customer's list of the BO. However, we have currently have no way to see if a customer is actually a guest unless you open their customer file to see the detail of the customer. I think adding a way of seeing guests in the customers list could be a good addition, wdyt ?

I think that most of my points are cause for another PR (except maybe the last one that's purely BO related), but since they are related to your PR, I would be interested to have your opinion on the matter.

Hlavtox commented:
Hi @eshraw and @Julievrz

> The concept of guest account is problematic from a semantic point of view. As a guest you don't expect to be able to transform a "guest account" into a "customer account" because you don't expect to have an account. If you decided to check-out as a guest you don't expect data persistence nor the possibility to mutate a "guest account", you would expect to have to sign up some way or another.

This concept is in prestashop for years and I didn't hear any customer complaining about anything. Don't worry, it's fine.

> You don't expect data persistence

Well, the data must be saved in some way. :-)

> nor the possibility to mutate a "guest account"

Yet the customer has it. Less info for him to fill in!

> In the guest order tracking page the guest is suggested to transform a "guest account" into a "customer account" by setting a password --> this differs from the check-out convert attempt and promotes this concept of "guest account" that we would probably like to avoid since it doesn't make much sense (semantically speaking).

Yes, because we already have all the required data in his "private zone" (guest tracking/order confirmation), so the only required data to register him is the password. No point of filling it again. I recently unified and fixed this in one of my PRs, quite a bit of work - https://github.com/PrestaShop/PrestaShop/pull/25395. Check it out. It was a mess before it.

> While I understand that technically, a guest in PrestaShop is a customer account with is_guest == true, semantically and functionally we need to steer away from that because it creates a situation in which it implies that we are storing guest checkout data into an account they didn't ask for. I stand by @Julievrz when she propose that we use "Create a guest".

- We already tried to improve it by using a wording `New customer` in hummingbird checkout. 
- Because everyone who buys in our store is a `Customer` > They are shown in a `Customers` section in backoffice.
- I agree that you can clarify it a bit more and think of new terms to distinguish them, but I think you are looking for problems where they aren't. `Guest customer` and `Registered customer`, that's it, nobody complains.

> In the FO, we need to be more consistent on guest-to-customer conversion, I believe we can be smart with our wordings here and use something along the lines of "Re-use your order informations to create an account" + password input on both pages allowing this conversion.

Good idea, I like this wording for customer conversion!

> The last point I'd like to discuss is how you would display guests in the customer's list ? As you mentionned, guests are customers with is_guest == true, thus they appear in the customer's list of the BO. However, we have currently have no way to see if a customer is actually a guest unless you open their customer file to see the detail of the customer. I think adding a way of seeing guests in the customers list could be a good addition, wdyt ?

I already worked on this, check these out:
- Displaying customer group when making new order - https://github.com/PrestaShop/PrestaShop/pull/25091
- Displaying customer group on order detail page - https://github.com/PrestaShop/PrestaShop/pull/25080
- Displaying customer group on customer listing page - https://github.com/PrestaShop/PrestaShop/pull/28183

eshraw commented:
Hey @Hlavtox , thank you for your detailled answer and sorry for the delay.

- Regarding the concept of guest account : the lack of people complaining about it doesn't necessarily means that there is no issue or opportunity related to this concept. If we take a look at it from another angle, people that would complain about this concept / wording would also be the one either not buying from the shop or buy from it only once. Either way, these most likely wouldn't be the customers contacting the merchants about it as they would probably avoid the shop altogether. Unfortunately, we can only make assumptions regarding both point of views as we lack data here :/ 
In any case, even if there truly is no one taking issue with it, it doesn't mean we can't improve it and transform something neutral in something positive for the merchants, I think we can both agree on that right ?
- Data persistence regarding the guest account concept : it has to be saved in some way indeed. But the how matters, as a "guest" I don't expected to have an "account", even if I know my name, address, etc. will be stored somehow. Mixing both seems odd to me as a customer even if I understand the value of it from an insider point of view (easier conversion from guest to customer account is a great thing)
- Regarding mutation from "guest account" to "customer account" : it's a good feature, there is no denying it, but tying it to the "guest account" makes it odd from a customer point of view as "I haven't registered yet, why do I have an account ? 🤔 ". With the ideas stemming from our exchange I think we can enhance this feature (cf next points)
- "Guest customer" & "Registered customer" : Yes, good idea 👍 . I understand where you are coming from, I see this more as an opportunity to improve an odd detail about PS to enhance the underlying feature (guest-to-customer conversion) rather than a problem that must be solved. Although in this case there is a bit of an industry standard alignement at play regarding the semantic behind "guest customer" and "guest account". In the end, I see this as tackling 2 birds with 1 stone, we transform an odd wording to something customers are more accustomed to and we improve an already good feature. :) 
- [#25395](https://github.com/PrestaShop/PrestaShop/pull/27769#:~:text=bit%20of%20work%20%2D-,%2325395,-.%20Check%20it%20out) : I understand better, thanks for the details 👍 . What do you think of pairing this with the wording I proposed "Re-use your order informations to create an account" to make it clearer as to why we don't ask for a refill of the infos ?
- Issues linked to the BO : Thanks for the links !

My take on our exchange :
- Use the wordings "Guest customer" & "Registered customer" as both are indeed customers
- Drop the "guest account" wording in favor of "guest customer"
- User something along the lines of "Re-use your order informations to create an account" to improve our guest-to-registered conversion feature (I believe this one would be another issue/PR unrelated to this one)
Wdyt ? @Hlavtox @Julievrz

Hlavtox committed:
Minor tweaks and CS fix

l-delin commented on src/PrestaShopBundle/Form/Admin/Sell/Customer/CustomerType.php:
```diff
@@ -123,6 +133,17 @@ public function buildForm(FormBuilderInterface $builder, array $options)
+        if ($options['show_guest_field'] === true) {
+            $builder
+                ->add('is_guest', SwitchType::class, [
+                    'label' => $this->trans('Register as guest', 'Admin.Global'),
```
```suggestion
                    'label' => $this->trans('Guest account', 'Admin.Global'),
```

Hlavtox replied:
Would also work, but I am thinking that in the future, we could probably also display this checkbox in the "edit" form, so maybe we should avoid the words `Register` and `Create`.

So what about `Guest customer` or `Guest account` or just `Guest`?

kpodemski replied:
IMHO it should be just `Guest account`

l-delin replied:
Let's stick with `Guest account` then. It seems clear enough to me. I'll edit my suggestion

l-delin commented on src/PrestaShopBundle/Form/Admin/Sell/Customer/CustomerType.php:
```diff
@@ -123,6 +133,17 @@ public function buildForm(FormBuilderInterface $builder, array $options)
+                ->add('is_guest', SwitchType::class, [
+                    'label' => $this->trans('Register as guest', 'Admin.Global'),
+                    'help' => $this->trans(
+                        'Adds a one-time guest customer with no password.',
```
```suggestion
                        'The user won't need a password to sign in to your store unless they convert their guest account into a regular one.',
```

l-delin replied:
wdyt @Hlavtox ? I edited the wording for clarity.

Hlavtox replied:
What about this, it's shorter.

`Guests don't have a password and can't log in to your store, unless they are converted to a regular customer.`

l-delin replied:
I think we should keep the wording I suggested, a double negation isn't great in a wording. 
Besides, I believe using "guests (...) are converted" is vague. (You don't "convert" people? 🧐). This help text needs to be as clear as possible. I will edit my suggestion, let me know what you think.

Hlavtox replied:
@l-delin I don't like it, because:
- We don't call customers users, users are if anywhere, in the backoffice.
- `Won't need a password to sign in` - nonsense, they **don't sign in** at all, they `don't have` the password.
- I checked it and we should use "transform" wording - that's used in both FO and BO.

What about this:
`Guests don't have a password and can't log into your store, unless their account is transformed to a customer account.`

l-delin replied:
Got it. Here's another proposition. It's shorter and clearer:

`Visitors can order as guests. At any time, you can turn your guests into customers.`

kpodemski replied:
@l-delin I like this version, but I would keep the educational part where we explain what it means to be a Guest.

l-delin replied:
`Visitors can order as guests without creating an account. At any time, you can turn your guests into customers.` ?

kpodemski replied:
LGTM :) wdyt @Hlavtox ?

kpodemski replied:
@l-delin now that I think more about it, there might be confusion because this information is visible in the back office, and there's already a group called "Visitors", etc. I think it's good to inform merchants about the actual consequences of adding a guest account, and the truth is that they won't have the password set, and they cannot log in as a regular customer does.

Hlavtox replied:
@l-delin Sorry Lea but the wording make no sense. I am using Prestashop for 8 years now, fixed the guest ordering, guest transformation, guest groups and now the guest creation and the wording just doesnt say what it should.

Why the hell should I see `Visitors can order as guests without creating an account. At any time, you can turn your guests into customers.` when I create a guest in a backoffice? Guests aren't visitors, they cannot order.

The wording should say that the Guest doesnt have a password, that it's a one time customer for single use and it can be later transformed.

`Guests don't have a password and can't log into your store, unless their account is transformed to a customer account.` says it all and it's short.

Hlavtox replied:
UPDATE: We have come with a good wording with @kpodemski. Our proposal, it's longer, but so far the best:

`Guests are quickly made customers with no password, who don't have access to the privileges of the registered ones. You can create as many guests as needed using the same email. This feature could be helpful if you take phone call orders and insert them into the system.`

or shorter version

`A quick customer with no password, who doesn't have access to the privileges of the registered one. You can create as many guests as needed using the same email. It could be helpful if you take phone call orders.`

l-delin replied:
Alright. Then, let's go with the shorter version. If there needs to be any precision, it'll be added to the doc when the feature is released. The shorter version is better as this is a help text and we don't want it to be too long.

So, an edited version (so it's shorter):
`Quick customers with no password, who don't have access to the privileges of registered ones. You can create as many guests as needed using the same email. It could be helpful if you take phone call orders.`

kpodemski commented on src/PrestaShopBundle/Form/Admin/Sell/Customer/CustomerType.php:
```diff
@@ -358,6 +383,20 @@ public function buildForm(FormBuilderInterface $builder, array $options)
         }
+
+        // We add a listener that will make password field not required, if we want to create a guest
+        $builder->addEventListener(FormEvents::PRE_SUBMIT, function (FormEvent $event) {
```
If you are motivated, you can separate this into another class, here's an example:
https://github.com/PrestaShop/PrestaShop/pull/31568/files

I'm sure @zuk3975 would appreciate it :D

Hlavtox replied:
Maybe next time, I will do something useful instead 🥲

kpodemski reviewed (dismissed):
One non-blocking comment, apart from that, LGTM. Well done 👏🏻

Hlavtox committed:
Finalize wording

Hlavtox commented:
@l-delin Wording addressed as per our discussions! :-)

l-delin commented:
Thanks!

kpodemski reviewed (dismissed):
Reapproving.

HanaRebaiQA reviewed (commented):
Hello @Hlavtox 

I have checked your PR. I found these issues : 

**Issue 1:** The toggle title should be `Registered as a Guest` as mentioned on the specifications

**Specifications :** 

![image](https://user-images.githubusercontent.com/92912932/223741969-1bfa8139-3836-4bec-bafe-8de8c30fe2c8.png)

![image](https://user-images.githubusercontent.com/92912932/223741995-8bee709b-bfe9-4898-ba55-6451c5b89e62.png)

**With PR :** 

![image](https://user-images.githubusercontent.com/92912932/223741938-95897f14-0a12-4469-964f-f0b7bcb4c7ac.png)

**Issue 2 :** When the toggle if the registered as a guest is enabled, the default group is set at Guest, and then the Select all should not be enabled 

**Specifications :** 

![image](https://user-images.githubusercontent.com/92912932/223742698-5a949aae-a520-41bf-9129-f9c0da4ea31e.png)

**With PR :** 

![image](https://user-images.githubusercontent.com/92912932/223742674-56c40c40-1a2d-4cb5-976f-a63ddd668d60.png)

**Issue 3 :** The password should not be a required field and the Enabled toggle should be disabled when editing the created guest.

1. Create a new customer
2. Enable the toggle to register as a guest
3. The password is not enabled : OK and the Enabled toggle is disabled : OK
4. Click on Save
5. Edit the created guest
6. See that these fields are not OK
7. Click on Save
8. See error

![image](https://user-images.githubusercontent.com/92912932/223743299-60aef059-7ed6-4786-85ca-0adc82d98106.png)


https://user-images.githubusercontent.com/92912932/223743481-7d848e47-2a92-4021-8cab-b91b162e2775.mp4

Could you please check it ?

Thank You!

Hlavtox committed:
Use NotBlank constraint only when creating a customer

Hlavtox commented:
Hi @HanaRebaiQA
> Issue 1: The toggle title should be Registered as a Guest as mentioned on the specifications

The specs have changed, I have prepared some initial wording and it has been finalized by Lea in this discussion. https://github.com/PrestaShop/PrestaShop/pull/27769#discussion_r1121381822

> Issue 2 : When the toggle if the registered as a guest is enabled, the default group is set at Guest, and then the Select all should not be enabled

The checkbox just remembers the last state, there are no specifications for it. When you check all of them, it's still off. Don't make things up. :smile:
![CPT2303081850-479x245](https://user-images.githubusercontent.com/6097524/223790869-d8e8fabe-1080-4ef9-b9e4-53a155a635f7.gif)

> Issue 3 : Enabled toggle should be disabled when editing the created guest.

Again, I am only modifying customer creation, guest edit stayed the same for now, so merchants can edit or fix previously broken guests. I will focus on the edit page sometime in the future, but I need to think it through.

> Issue 3 : The password should not be a required field when editing the created guest.

Good catch, this is an issue. I added a proper check for not empty field but forgot to check the edit scenario. I fixed it in my last commit. Thank you. 👍

Hlavtox commented:
### Offtopic for reviewers of my last commit

The code validating the password is messy, but it's just moved from the middle of the form. It's using password policy constraints for JS validation and password value object policy for validation on PHP side. So if you set minimum password length to 1, it will let you submit it, but it will fail and say that the minimum length is 5.

Even if I fixed it in the form and used `$minLength` instead of `Password::MIN_LENGTH`, it could later fail when constructing the `Password` value object, so it would not help at all. Better not mix this topic into this PR at all. I will fix it in this separate PR properly - https://github.com/PrestaShop/PrestaShop/pull/29776, after this is merged. 👍

matks reviewed (changes requested):
@Hlavtox I checked the last commit and it is valid but your PR needs a rebase as file `src/PrestaShopBundle/Controller/Admin/Sell/Customer/CustomerController.php` is marked in conflict

florine2623 reviewed (approved):
Hello @Hlavtox ,

Works as expected 👍 
Thanks! It is QA ✅

Hlavtox commented:
Thank you @florine2623! :-)
